force_wfmanager/server/zmq_server.py

```diff
@@ -127,23 +201,47 @@ def stop(self):
             # ignore
             log.exception("Error while attempting to stop server")
         finally:
-            socket.close()
+            if socket:
+                socket.close()
 
     def _setup_sockets(self):
         """Sets up the sockets."""
         context = self._context
         pub_socket = context.socket(zmq.SUB)
         pub_socket.setsockopt(zmq.SUBSCRIBE, "".encode("utf-8"))
         pub_socket.setsockopt(zmq.LINGER, 0)
-        pub_socket.bind(self.config.pub_url)
+        pub_port = pub_socket.bind_to_random_port("tcp://*")
 
         sync_socket = context.socket(zmq.REP)
         sync_socket.setsockopt(zmq.LINGER, 0)
-        sync_socket.bind(self.config.sync_url)
+        sync_port = sync_socket.bind_to_random_port("tcp://*")
 
         inproc_socket = context.socket(zmq.PAIR)
         inproc_socket.bind("inproc://stop")
-        return pub_socket, sync_socket, inproc_socket
+        return pub_socket, pub_port, sync_socket, sync_port, inproc_socket
+
+    def _close_network_sockets_noerror(self):
```

Comment:
Just a naming thing, close_network_sockets_noerror is used for both closing sockets normally and after an error. 
I suppose now there's a distinct on_error_callback we could have a method for each case, but it might be easier just to rename it close_network_sockets and keep things as they are at the minute.

Reply:
it is mostly referred to the fact that the method cannot raise. The close operation is quite delicate, as it must be done in any case, and can fail if there was a failure. I will call it noexc instead.